packages/network-controller/src/NetworkController.ts

```diff
@@ -429,6 +455,77 @@ export class NetworkController extends BaseControllerV2<
     }
     return Promise.resolve(true);
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param config - The network configuration.
+   * @param config.rpcUrl - The network configuration RPC URL.
+   * @param config.chainId - The chain ID of the network, as per EIP-155.
+   * @param config.ticker - Currency ticker.
+   * @param config.nickname - Personalized network name.
+   * @param config.rpcPrefs - Personalized preferences.
+   * @returns uuid for the added or updated network configuration
+   */
+  upsertNetworkConfiguration({
+    rpcUrl,
+    chainId,
+    ticker,
+    nickname,
+    rpcPrefs,
+  }: NetworkConfiguration): string {
+    const networkConfigurations = { ...this.state.networkConfigurations };
+
+    const newNetworkConfiguration: NetworkConfiguration = {
+      rpcUrl,
+      chainId,
+      ticker,
+      nickname,
+      rpcPrefs,
+    };
+
+    for (const configUUID in networkConfigurations) {
+      if (
+        networkConfigurations[configUUID].rpcUrl.toLowerCase() ===
+        rpcUrl.toLowerCase()
+      ) {
+        // update existing config
+        networkConfigurations[configUUID] = newNetworkConfiguration;
+        this.update((state) => {
+          state.networkConfigurations = networkConfigurations;
+        });
+        return configUUID;
+      }
+    }
+
+    const uuid = random();
+    networkConfigurations[uuid] = newNetworkConfiguration;
+
+    this.update((state) => {
+      state.networkConfigurations = networkConfigurations;
+    });
+
+    return uuid;
+  }
+
+  /**
+   * Removes network configuration from state.
+   *
+   * @param uuid - Custom RPC URL.
+   */
+  removeNetworkConfigurations(uuid: string) {
+    const networkConfigurations = { ...this.state.networkConfigurations };
+    delete networkConfigurations[uuid];
+
+    this.update((state) => {
+      state.networkConfigurations = networkConfigurations;
+    });
```

Comment:
This is a similar point as above — this will update state directly. What about:

``` suggestion
    this.update((state) => {
      delete state.networkConfigurations[uuid];
    });
```

Reply:
updated [here](https://github.com/MetaMask/core/pull/1064/commits/ac392e90922842cc69a7cb6a7377557c532dbc35)